stdlib/public/Backtracing/MemoryReader.swift

```diff
@@ -60,55 +76,174 @@ extension MemoryReader {
-                                          UInt64(address),
-                                          UInt64(size),
-                                          buffer.baseAddress,
-                                          &sizeOut)
-
-    if result != _SWIFT_KERN_SUCCESS {
+    let result = mach_vm_read_overwrite(task,
+                                        UInt64(address),
+                                        UInt64(size),
+                                        mach_vm_address_t(
+                                          Int(bitPattern: buffer.baseAddress)
+                                        ),
+                                        &sizeOut)
+
+    if result != KERN_SUCCESS {
       throw MachError(result: result)
     }
   }
 }
 
 @_spi(MemoryReaders) public struct LocalMemoryReader: MemoryReader {
   public typealias Address = UInt64
+  public typealias Size = UInt64
+
+  public func fetch<T>(from address: Address,
+                       into buffer: UnsafeMutableBufferPointer<T>) throws {
+    let reader = RemoteMemoryReader(task: mach_task_self())
+    return try reader.fetch(from: address, into: buffer)
+  }
+}
+#endif
+
+#if os(Linux)
+@_spi(MemoryReaders) public struct POSIXError: Error {
+  var errno: CInt
+}
+
+@_spi(MemoryReaders) public struct MemserverError: Error {
+  var message: String
+}
+
+@_spi(MemoryReaders) public struct MemserverMemoryReader: MemoryReader {
+  private var fd: CInt
+
+  public init(fd: CInt) {
+    self.fd = fd
+  }
+
+  private func sendRequest(for bytes: Size, from addr: Address) throws {
+    var request = memserver_req(addr: addr, len: bytes)
+    try withUnsafeBytes(of: &request){ ptr in
+      let ret = write(fd, ptr.baseAddress, ptr.count)
+      if ret < 0 || ret != ptr.count {
+        throw POSIXError(errno: _swift_get_errno())
+      }
+    }
+  }
+
+  private func receiveReply() throws -> memserver_resp {
+    var response = memserver_resp(addr: 0, len: 0)
+    try withUnsafeMutableBytes(of: &response){ ptr in
+      let ret = read(fd, ptr.baseAddress, ptr.count)
+      if ret < 0 || ret != ptr.count {
+        throw POSIXError(errno: _swift_get_errno())
+      }
+    }
+    return response
+  }
+
+  public func fetch<T>(from addr: Address,
+                       into buffer: UnsafeMutableBufferPointer<T>) throws {
+    try buffer.withMemoryRebound(to: UInt8.self) { bytes in
+      try sendRequest(for: Size(bytes.count), from: addr)
+
+      var done = 0
+      while done < bytes.count {
+        let reply = try receiveReply()
+
+        if reply.len < 0 {
+          throw MemserverError(message: "Unreadable at \(hex(addr))")
+        }
+
+        if done + Int(reply.len) > bytes.count {
+          throw MemserverError(message: "Overrun at \(hex(addr)) trying to read \(bytes.count) bytes")
+        }
+
+        let ret = read(fd,
+                       bytes.baseAddress!.advanced(by: done),
+                       Int(reply.len))
+
+        if ret < 0 || ret != reply.len {
```

Comment:
Another potential partial read.

Reply:
Yeah, while I haven't seen it happen, I think you're right that I should handle that.